src/Type/EntityStorageDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,94 @@
+<?php declare(strict_types=1);
+
+namespace mglaman\PHPStanDrupal\Type;
+
+use Drupal\Core\Entity\EntityStorageInterface;
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\TypeCombinator;
+
+class EntityStorageDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
+{
+    /**
+     * @var string[]
+     */
+    private $entityTypeStorageMapping;
+
+    /**
+     * @var string[]
+     */
+    private $entityStorageMapping;
+
+    /**
+     * EntityTypeManagerGetStorageDynamicReturnTypeExtension constructor.
+     *
+     * @param string[] $entityTypeStorageMapping
+     */
+    public function __construct(
+        array $entityTypeStorageMapping = [],
+        array $entityStorageMapping = []
+    ) {
+        $this->entityTypeStorageMapping = $entityTypeStorageMapping;
+        $this->entityStorageMapping = $entityStorageMapping;
+    }
+
+    public function getClass(): string
+    {
+        return EntityStorageInterface::class;
+    }
+
+    public function isMethodSupported(MethodReflection $methodReflection): bool
+    {
+        return \in_array(
+            $methodReflection->getName(),
+            [
+                'create',
+                'load',
+                'loadMultiple',
+                'loadByProperties',
+                'loadUnchanged',
+            ],
+            true
+        );
+    }
+
+    public function getTypeFromMethodCall(
+        MethodReflection $methodReflection,
+        MethodCall $methodCall,
+        Scope $scope
+    ): \PHPStan\Type\Type {
+        $callerType = $scope->getType($methodCall->var);
+
+        if (!$callerType instanceof ObjectType) {
+            return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+        }
+
+        $entityTypeStorageMapping = \array_flip($this->entityTypeStorageMapping);
+        $entityType = $entityTypeStorageMapping[$callerType->getClassName()] ?? null;
+        if (null === $entityType) {
+            return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+        }
+
```

Comment:
I don't think we need to check if the entity storage was mapped, at all. If we don't have the mapping, it's perfectly fine for us to trust the default storage interface BUT define the entity type class returned.

Reply:
This part has been removed.